staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -83,30 +84,88 @@ type OIDCOptions struct {
 	// GroupsPrefix, if specified, causes claims mapping to group names to be prefixed with the
 	// value. A value "oidc:" would result in groups like "oidc:engineering" and "oidc:marketing".
 	GroupsPrefix string
+
+	// SupportedSigningAlgs sets the accepted set of JOSE signing algorithms that
+	// can be used by the provider to sign tokens.
+	//
+	// https://tools.ietf.org/html/rfc7518#section-3.1
+	//
+	// This value defaults to RS256, the value recommended by the OpenID Connect
+	// spec:
+	//
+	// https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation
+	SupportedSigningAlgs []string
+
+	// now is used for testing. It defaults to time.Now.
+	now func() time.Time
 }
 
-type OIDCAuthenticator struct {
+type Authenticator struct {
 	issuerURL string
 
-	trustedClientID string
-
 	usernameClaim  string
 	usernamePrefix string
 	groupsClaim    string
 	groupsPrefix   string
 
-	httpClient *http.Client
+	// Contains an *oidc.IDTokenVerifier. Do not access directly use the
+	// idTokenVerifier method.
+	verifier atomic.Value
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (a *Authenticator) setVerifier(v *oidc.IDTokenVerifier) {
+	a.verifier.Store(v)
+}
 
-	// Contains an *oidc.Client. Do not access directly. Use client() method.
-	oidcClient atomic.Value
+func (a *Authenticator) idTokenVerifier() (*oidc.IDTokenVerifier, bool) {
+	if v := a.verifier.Load(); v != nil {
+		return v.(*oidc.IDTokenVerifier), true
+	}
+	return nil, false
+}
 
-	// Guards the close method and is used to lock during initialization and closing.
-	mu    sync.Mutex
-	close func() // May be nil
+func (a *Authenticator) Close() {
+	if a.cancel != nil {
+		a.cancel()
+	}
 }
 
-// New creates a token authenticator which validates OpenID Connect ID Tokens.
-func New(opts OIDCOptions) (*OIDCAuthenticator, error) {
+func New(opts OIDCOptions) (*Authenticator, error) {
+	return newAuthenticator(opts, func(a *Authenticator, config *oidc.Config) {
+		// Asynchronously attempt to initialize the authenticator. This enables
+		// self-hosted providers, providers that run on top of Kubernetes itself.
+		go wait.PollUntil(time.Second*10, func() (done bool, err error) {
+			provider, err := oidc.NewProvider(a.ctx, a.issuerURL)
+			if err != nil {
+				glog.Errorf("oidc authenticator: initializing plugin: %v", err)
+				return false, nil
+			}
+
+			verifier := provider.Verifier(config)
+			a.setVerifier(verifier)
+			return true, nil
+		}, a.ctx.Done())
+	})
+}
+
+// whilelist of signing algorithms to ensure users don't mistakenly pass something
+// goofy.
```

Comment:
don't want to enable `none`? :)

Reply:
I swear I remember someone accidentally using tokens with `HS256`.